File: foundation/interfaces/connection.py
"""
PeerMesh Foundation - Connection Abstraction Interface

This file defines the Python protocols (interfaces) for the connection abstraction system.
Modules declare what connections they need (database, cache, etc.) and the foundation
resolves these to available providers at runtime.

Core Principle: Modules declare requirements, not implementations.
A module that needs "a database" can work with Postgres, MySQL, or SQLite - the
foundation matches requirements to available providers.

Usage:
    from foundation.interfaces.connection import (
        ConnectionRequirement,
        ConnectionResolver,
        ResolvedConnection,
    )

Example:
    requirement = ConnectionRequirement(
        type='database',
        providers=['postgres', 'mysql', 'sqlite'],
        required=True,
        name='primary-db'
    )

    result = await resolver.resolve('my-module', [requirement])
    if result.success:
        for conn in result.resolved:
            print(f"{conn.requirement_name} -> {conn.provider_name}")
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Literal, Optional, Protocol, TypeVar, Union


class ConnectionType(str, Enum):
    """Connection types supported by the foundation."""
    DATABASE = 'database'
    CACHE = 'cache'
    STORAGE = 'storage'
    QUEUE = 'queue'
    EVENTBUS = 'eventbus'
    CUSTOM = 'custom'


class DatabaseProvider(str, Enum):
    """Standard database providers."""
    POSTGRES = 'postgres'
    MYSQL = 'mysql'
    MARIADB = 'mariadb'
    SQLITE = 'sqlite'
    MONGODB = 'mongodb'
    COCKROACHDB = 'cockroachdb'


class CacheProvider(str, Enum):
    """Standard cache providers."""
    REDIS = 'redis'
    MEMCACHED = 'memcached'
    VALKEY = 'valkey'
    DRAGONFLY = 'dragonfly'


class StorageProvider(str, Enum):
    """Standard storage providers."""
    LOCAL = 'local'
    S3 = 's3'
    MINIO = 'minio'
    GCS = 'gcs'
    AZURE_BLOB = 'azure-blob'


class QueueProvider(str, Enum):
    """Standard queue providers."""
    RABBITMQ = 'rabbitmq'
    REDIS = 'redis'
    NATS = 'nats'
    KAFKA = 'kafka'
    SQS = 'sqs'


class EventBusProvider(str, Enum):
    """Standard event bus providers."""
    REDIS = 'redis'
    NATS = 'nats'
    KAFKA = 'kafka'
    MEMORY = 'memory'
    NOOP = 'noop'


@dataclass
class ConnectionConfig:
    """
    Configuration options for a connection.

    Attributes:
        host: Hostname or IP address
        port: Port number
        database: Database name (for database connections)
        username: Username for authentication
        password: Password for authentication (typically from env or secrets)
        ssl: Whether to use SSL/TLS
        pool_size: Connection pool size
        timeout: Connection timeout in milliseconds
        extra: Provider-specific configuration options
    """
    host: Optional[str] = None
    port: Optional[int] = None
    database: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None
    ssl: bool = False
    pool_size: int = 10
    timeout: int = 30000
    extra: Dict[str, Any] = field(default_factory=dict)


@dataclass
class ConnectionRequirement:
    """
    A connection requirement declared by a module.

    Modules declare what they need, not which specific implementation.
    The foundation matches requirements to available providers.

    Attributes:
        type: Type of connection required
        providers: Acceptable providers for this connection (in order of preference)
        required: Whether this connection is required for the module to function
        name: Unique name for this connection requirement
        config: Default configuration for this connection

    Example:
        requirement = ConnectionRequirement(
            type=ConnectionType.DATABASE,
            providers=['postgres', 'mysql', 'sqlite'],
            required=True,
            name='primary-db',
            config=ConnectionConfig(pool_size=20)
        )
    """
    type: ConnectionType
    providers: List[str]
    required: bool = True
    name: Optional[str] = None
    config: Optional[ConnectionConfig] = None


@dataclass
class ConnectionProvider:
    """
    A connection provider offered by a module.

    Provider modules (like postgres-provider, redis-provider) declare what
    connection types they can satisfy.

    Attributes:
        name: Provider name (e.g., 'postgres', 'redis')
        type: Type of connection this provider offers
        provides: List of provider identifiers this satisfies
        version: Version of the provider software
        config: Default configuration for this provider

    Example:
        provider = ConnectionProvider(
            name='postgres',
            type=ConnectionType.DATABASE,
            provides=['postgres', 'postgresql'],
            version='16',
            config=ConnectionConfig(host='postgres', port=5432)
        )
    """
    name: str
    type: ConnectionType
    provides: List[str]
    version: Optional[str] = None
    config: Optional[ConnectionConfig] = None


@dataclass
class ResolvedConnection:
    """
    A connection that has been matched to a provider.

    After resolution, modules receive this object containing the final
    configuration to connect to the service.

    Attributes:
        requirement_name: Name of the requirement that was satisfied
        provider_module: Module ID providing this connection
        provider_name: Name of the provider being used
        type: Type of the connection
        config: Final resolved configuration
        connection_string: Connection string (if applicable)
        env_vars: Environment variables to set for this connection
    """
    requirement_name: str
    provider_module: str
    provider_name: str
    type: ConnectionType
    config: ConnectionConfig
    connection_string: Optional[str] = None
    env_vars: Optional[Dict[str, str]] = None


@dataclass
class UnresolvedConnection:
    """
    Information about an unresolved connection requirement.

    Attributes:
        requirement: The requirement that could not be satisfied
        reason: Reason the connection could not be resolved
    """
    requirement: ConnectionRequirement
    reason: str


@dataclass
class ResolutionResult:
    """
    Result of attempting to resolve module connections.

    Attributes:
        success: Whether all required connections were resolved
        resolved: Successfully resolved connections
        unresolved: Connections that could not be resolved
        warnings: Non-fatal warnings during resolution
    """
    success: bool
    resolved: List[ResolvedConnection]
    unresolved: List[UnresolvedConnection] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)


T = TypeVar('T')


class Connection(Protocol):
    """
    Connection represents an active connection to a service.

    This is the runtime interface that modules use to interact with
    their resolved connections.
    """

    @property
    def info(self) -> ResolvedConnection:
        """The resolved connection information."""
        ...

    async def ping(self) -> bool:
        """Check if the connection is alive."""
        ...

    def get_client(self) -> Any:
        """Get the raw connection client (type depends on provider)."""
        ...

    async def close(self) -> None:
        """Close the connection."""
        ...


class ConnectionFactory(Protocol):
    """
    ConnectionFactory creates Connection instances from resolved configurations.

    Each provider module registers a factory for its connection type.
    """

    @property
    def provider_name(self) -> str:
        """Provider name this factory handles."""
        ...

    @property
    def connection_type(self) -> ConnectionType:
        """Connection type this factory creates."""
        ...

    async def create(self, resolved: ResolvedConnection) -> Connection:
        """
        Create a connection from resolved configuration.

        Args:
            resolved: Resolved connection configuration

        Returns:
            Active connection instance
        """
        ...


class ConnectionResolver(Protocol):
    """
    ConnectionResolver resolves module connection requirements to available providers.

    The resolver:
    1. Reads module's connection requirements
    2. Scans available provider modules
    3. Matches requirements to providers based on type and provider list
    4. Returns resolved configurations or errors

    Example:
        result = await resolver.resolve('my-module', requirements)
        if result.success:
            for conn in result.resolved:
                print(f"{conn.requirement_name} -> {conn.provider_name}")
        else:
            for unresolved in result.unresolved:
                print(f"Cannot satisfy: {unresolved.requirement.name}")
    """

    async def resolve(
        self,
        module_id: str,
        requirements: List[ConnectionRequirement],
    ) -> ResolutionResult:
        """
        Resolve connection requirements for a module.

        Args:
            module_id: ID of the module requesting connections
            requirements: Connection requirements to resolve

        Returns:
            Resolution result with resolved and unresolved connections
        """
        ...

    async def get_available_providers(self) -> Dict[str, ConnectionProvider]:
        """
        Get all available connection providers.

        Returns:
            Dictionary of provider names to provider definitions
        """
        ...

    async def is_provider_available(self, provider_name: str) -> bool:
        """
        Check if a specific provider is available.

        Args:
            provider_name: Name of the provider to check

        Returns:
            True if the provider is installed and available
        """
        ...

    async def get_connection(
        self,
        module_id: str,
        connection_name: str,
    ) -> Optional[ResolvedConnection]:
        """
        Get connection information for a resolved connection.

        Args:
            module_id: Module that owns the connection
            connection_name: Name of the connection

        Returns:
            Resolved connection info or None if not found
        """
        ...


def build_connection_string(
    conn_type: ConnectionType,
    provider: str,
    config: ConnectionConfig,
) -> str:
    """
    Build a connection string from configuration.

    Args:
        conn_type: Connection type
        provider: Provider name
        config: Connection configuration

    Returns:
        Connection string (format depends on type/provider)
    """
    host = config.host or 'localhost'
    port = config.port
    database = config.database
    username = config.username
    password = config.password
    ssl = config.ssl

    # Build auth part
    auth = ''
    if username:
        auth = username
        if password:
            auth = f"{auth}:{password}"
        auth = f"{auth}@"

    # Build port part
    port_str = f":{port}" if port else ""

    if conn_type == ConnectionType.DATABASE:
        if provider in ('postgres', 'postgresql'):
            db = database or 'postgres'
            ssl_param = '?sslmode=require' if ssl else ''
            return f"postgresql://{auth}{host}{port_str}/{db}{ssl_param}"
        elif provider in ('mysql', 'mariadb'):
            db = database or 'mysql'
            return f"mysql://{auth}{host}{port_str}/{db}"
        elif provider == 'mongodb':
            db = database or 'test'
            return f"mongodb://{auth}{host}{port_str}/{db}"
        elif provider == 'sqlite':
            return database or ':memory:'
        else:
            return f"{provider}://{host}{port_str}"
    elif conn_type == ConnectionType.CACHE:
        return f"redis://{auth}{host}{port_str}/0"
    else:
        return f"{provider}://{host}{port_str}"

File: foundation/interfaces/connection.ts
/**
 * PeerMesh Foundation - Connection Abstraction Interface
 *
 * This file defines the TypeScript interfaces for the connection abstraction system.
 * Modules declare what connections they need (database, cache, etc.) and the foundation
 * resolves these to available providers at runtime.
 *
 * Core Principle: Modules declare requirements, not implementations.
 * A module that needs "a database" can work with Postgres, MySQL, or SQLite - the
 * foundation matches requirements to available providers.
 *
 * @module foundation/interfaces/connection
 * @version 1.0.0
 */

/**
 * Connection types supported by the foundation.
 */
export type ConnectionType = 'database' | 'cache' | 'storage' | 'queue' | 'eventbus' | 'custom';

/**
 * Standard database providers.
 */
export type DatabaseProvider = 'postgres' | 'mysql' | 'mariadb' | 'sqlite' | 'mongodb' | 'cockroachdb';

/**
 * Standard cache providers.
 */
export type CacheProvider = 'redis' | 'memcached' | 'valkey' | 'dragonfly';

/**
 * Standard storage providers.
 */
export type StorageProvider = 'local' | 's3' | 'minio' | 'gcs' | 'azure-blob';

/**
 * Standard queue providers.
 */
export type QueueProvider = 'rabbitmq' | 'redis' | 'nats' | 'kafka' | 'sqs';

/**
 * Standard event bus providers.
 */
export type EventBusProvider = 'redis' | 'nats' | 'kafka' | 'memory' | 'noop';

/**
 * All provider types.
 */
export type Provider = DatabaseProvider | CacheProvider | StorageProvider | QueueProvider | EventBusProvider | string;

/**
 * Configuration options for a connection.
 */
export interface ConnectionConfig {
  /** Hostname or IP address */
  host?: string;

  /** Port number */
  port?: number;

  /** Database name (for database connections) */
  database?: string;

  /** Username for authentication */
  username?: string;

  /** Password for authentication (typically from env or secrets) */
  password?: string;

  /** Whether to use SSL/TLS */
  ssl?: boolean;

  /** Connection pool size */
  poolSize?: number;

  /** Connection timeout in milliseconds */
  timeout?: number;

  /** Provider-specific configuration options */
  extra?: Record<string, unknown>;
}

/**
 * A connection requirement declared by a module.
 *
 * Modules declare what they need, not which specific implementation.
 * The foundation matches requirements to available providers.
 *
 * @example
 * ```typescript
 * const requirement: ConnectionRequirement = {
 *   type: 'database',
 *   providers: ['postgres', 'mysql', 'sqlite'], // Any of these will work
 *   required: true,
 *   name: 'primary-db',
 *   config: { poolSize: 20 }
 * };
 * ```
 */
export interface ConnectionRequirement {
  /** Type of connection required */
  type: ConnectionType;

  /** Acceptable providers for this connection (in order of preference) */
  providers: Provider[];

  /** Whether this connection is required for the module to function */
  required?: boolean;

  /** Unique name for this connection requirement */
  name?: string;

  /** Default configuration for this connection */
  config?: ConnectionConfig;
}

/**
 * A connection provider offered by a module.
 *
 * Provider modules (like postgres-provider, redis-provider) declare what
 * connection types they can satisfy.
 *
 * @example
 * ```typescript
 * const provider: ConnectionProvider = {
 *   name: 'postgres',
 *   type: 'database',
 *   provides: ['postgres', 'postgresql'],
 *   version: '16',
 *   config: { host: 'postgres', port: 5432 }
 * };
 * ```
 */
export interface ConnectionProvider {
  /** Provider name (e.g., 'postgres', 'redis') */
  name: string;

  /** Type of connection this provider offers */
  type: ConnectionType;

  /** List of provider identifiers this satisfies */
  provides: string[];

  /** Version of the provider software */
  version?: string;

  /** Default configuration for this provider */
  config?: ConnectionConfig;
}

/**
 * A connection that has been matched to a provider.
 *
 * After resolution, modules receive this object containing the final
 * configuration to connect to the service.
 */
export interface ResolvedConnection {
  /** Name of the requirement that was satisfied */
  requirementName: string;

  /** Module ID providing this connection */
  providerModule: string;

  /** Name of the provider being used */
  providerName: string;

  /** Type of the connection */
  type: ConnectionType;

  /** Final resolved configuration */
  config: ConnectionConfig;

  /** Connection string (if applicable) */
  connectionString?: string;

  /** Environment variables to set for this connection */
  envVars?: Record<string, string>;
}

/**
 * Information about an unresolved connection requirement.
 */
export interface UnresolvedConnection {
  /** The requirement that could not be satisfied */
  requirement: ConnectionRequirement;

  /** Reason the connection could not be resolved */
  reason: string;
}

/**
 * Result of attempting to resolve module connections.
 */
export interface ResolutionResult {
  /** Whether all required connections were resolved */
  success: boolean;

  /** Successfully resolved connections */
  resolved: ResolvedConnection[];

  /** Connections that could not be resolved */
  unresolved: UnresolvedConnection[];

  /** Non-fatal warnings during resolution */
  warnings: string[];
}

/**
 * ConnectionResolver resolves module connection requirements to available providers.
 *
 * The resolver:
 * 1. Reads module's connection requirements
 * 2. Scans available provider modules
 * 3. Matches requirements to providers based on type and provider list
 * 4. Returns resolved configurations or errors
 *
 * @example
 * ```typescript
 * const resolver: ConnectionResolver = getConnectionResolver();
 *
 * // Resolve connections for a module
 * const result = await resolver.resolve('my-module', requirements);
 *
 * if (result.success) {
 *   for (const conn of result.resolved) {
 *     console.log(`${conn.requirementName} -> ${conn.providerName}`);
 *   }
 * } else {
 *   for (const unresolved of result.unresolved) {
 *     console.error(`Cannot satisfy: ${unresolved.requirement.name}`);
 *   }
 * }
 * ```
 */
export interface ConnectionResolver {
  /**
   * Resolve connection requirements for a module.
   *
   * @param moduleId - ID of the module requesting connections
   * @param requirements - Connection requirements to resolve
   * @returns Resolution result with resolved and unresolved connections
   */
  resolve(moduleId: string, requirements: ConnectionRequirement[]): Promise<ResolutionResult>;

  /**
   * Get all available connection providers.
   *
   * @returns Map of provider names to provider definitions
   */
  getAvailableProviders(): Promise<Map<string, ConnectionProvider>>;

  /**
   * Check if a specific provider is available.
   *
   * @param providerName - Name of the provider to check
   * @returns True if the provider is installed and available
   */
  isProviderAvailable(providerName: string): Promise<boolean>;

  /**
   * Get connection information for a resolved connection.
   *
   * @param moduleId - Module that owns the connection
   * @param connectionName - Name of the connection
   * @returns Resolved connection info or undefined if not found
   */
  getConnection(moduleId: string, connectionName: string): Promise<ResolvedConnection | undefined>;
}

/**
 * Connection represents an active connection to a service.
 *
 * This is the runtime interface that modules use to interact with
 * their resolved connections.
 */
export interface Connection {
  /** The resolved connection information */
  info: ResolvedConnection;

  /** Check if the connection is alive */
  ping(): Promise<boolean>;

  /** Get the raw connection client (type depends on provider) */
  getClient<T = unknown>(): T;

  /** Close the connection */
  close(): Promise<void>;
}

/**
 * ConnectionFactory creates Connection instances from resolved configurations.
 *
 * Each provider module registers a factory for its connection type.
 */
export interface ConnectionFactory {
  /** Provider name this factory handles */
  providerName: string;

  /** Connection type this factory creates */
  connectionType: ConnectionType;

  /**
   * Create a connection from resolved configuration.
   *
   * @param resolved - Resolved connection configuration
   * @returns Active connection instance
   */
  create(resolved: ResolvedConnection): Promise<Connection>;
}

/**
 * Utility function to build a connection string from config.
 *
 * @param type - Connection type
 * @param provider - Provider name
 * @param config - Connection configuration
 * @returns Connection string (format depends on type/provider)
 */
export function buildConnectionString(
  type: ConnectionType,
  provider: string,
  config: ConnectionConfig
): string {
  const { host = 'localhost', port, database, username, password, ssl } = config;

  switch (type) {
    case 'database':
      switch (provider) {
        case 'postgres':
        case 'postgresql':
          return `postgresql://${username ? `${username}${password ? `:${password}` : ''}@` : ''}${host}${port ? `:${port}` : ''}/${database || 'postgres'}${ssl ? '?sslmode=require' : ''}`;
        case 'mysql':
        case 'mariadb':
          return `mysql://${username ? `${username}${password ? `:${password}` : ''}@` : ''}${host}${port ? `:${port}` : ''}/${database || 'mysql'}`;
        case 'mongodb':
          return `mongodb://${username ? `${username}${password ? `:${password}` : ''}@` : ''}${host}${port ? `:${port}` : ''}/${database || 'test'}`;
        case 'sqlite':
          return database || ':memory:';
        default:
          return `${provider}://${host}${port ? `:${port}` : ''}`;
      }
    case 'cache':
      return `redis://${username ? `${username}${password ? `:${password}` : ''}@` : ''}${host}${port ? `:${port}` : ''}/0`;
    default:
      return `${provider}://${host}${port ? `:${port}` : ''}`;
  }
}
